File: src/drone/src/GPS.cpp
#include <rclcpp/rclcpp.hpp>
#include <serial/serial.h>
#include <drone_interfaces/msg/sensor_data.hpp>
#include <string>
#include <sstream>
#include <cmath>
#include <unordered_map>
#include <memory>

using namespace std::chrono_literals;

class SerialReader : public rclcpp::Node
{
public:
    SerialReader()
    : Node("Sensor_node")
    {

        serial_port_ = "/dev/ttyUSB0"; 
        baud_rate_ = 115200;

        RCLCPP_INFO(this->get_logger(), "Starting Sensor node. Trying to connect to serial port %s.", serial_port_.c_str());
        sensor_publisher_ = this->create_publisher<drone_interfaces::msg::SensorData>("/sensor", 10);
        timer_ = this->create_wall_timer(5ms, std::bind(&SerialReader::timer_callback, this));
        
        try {
            serial_.setPort(serial_port_);
            serial_.setBaudrate(baud_rate_);
            serial::Timeout timeout = serial::Timeout::simpleTimeout(1000);
            serial_.setTimeout(timeout);
            serial_.open();
        } catch (serial::IOException &e) {
            RCLCPP_ERROR(this->get_logger(), "Unable to open port %s", serial_port_.c_str());
        }

        RCLCPP_INFO(this->get_logger(), "SerialReader node started, reading from %s at %d baud rate.",
                    serial_port_.c_str(), baud_rate_);
    }

private:
    void timer_callback()
    {
        if (serial_.available())
        {
            try {
                std::string line = serial_.readline();
                RCLCPP_INFO(this->get_logger(), "Read from serial: %s", line.c_str());

                auto msg = drone_interfaces::msg::SensorData();
                bool new_data = false;

                // Parse the GPS data from the string
                if (line.find("<gps>") != std::string::npos && line.find("</gps>") != std::string::npos)
                {
                    auto data = parse_gps_data(line);
                    if (data)
                    {
                        msg.lat = string_to_double(data->at("lat"));
                        msg.lon = string_to_double(data->at("lon"));
                        msg.time = data->at("time");
                        msg.speed = string_to_double(data->at("speed"));
                        msg.pdop = string_to_double(data->at("pdop"));
                        msg.hdop = string_to_double(data->at("hdop"));
                        new_data = true;
                    }
                }

                // Parse the AHRS data from the string
                if (line.find("<AHRS>") != std::string::npos && line.find("</AHRS>") != std::string::npos)
                {
                    auto data = parse_ahrs_data(line);
                    if (data)
                    {
                        if (data->count("acc_x")) msg.acc_x = string_to_double(data->at("acc_x"));
                        if (data->count("acc_y")) msg.acc_y = string_to_double(data->at("acc_y"));
                        if (data->count("acc_z")) msg.acc_z = string_to_double(data->at("acc_z"));
                        if (data->count("gyro_x")) msg.gyro_x = string_to_double(data->at("gyro_x"));
                        if (data->count("gyro_y")) msg.gyro_y = string_to_double(data->at("gyro_y"));
                        if (data->count("gyro_z")) msg.gyro_z = string_to_double(data->at("gyro_z"));
                        if (data->count("mag_x")) msg.mag_x = string_to_double(data->at("mag_x"));
                        if (data->count("mag_y")) msg.mag_y = string_to_double(data->at("mag_y"));
                        if (data->count("mag_z")) msg.mag_z = string_to_double(data->at("mag_z"));
                        if (data->count("altitude")) msg.altitude = string_to_double(data->at("altitude"));
                        new_data = true;
                    }
                }

                if (new_data)
                {
                    sensor_publisher_->publish(msg);
                }
            } catch (serial::IOException &e) {
                RCLCPP_ERROR(this->get_logger(), "Error reading from serial port: %s", e.what());
            }
        }
    }

    std::shared_ptr<std::unordered_map<std::string, std::string>> parse_gps_data(const std::string& data)
    {
        auto gps_data = std::make_shared<std::unordered_map<std::string, std::string>>();

        auto parse_tag_value = [&data](const std::string& tag) -> std::string {
            std::string start_tag = "<" + tag + ">";
            std::string end_tag = "</" + tag + ">";
            size_t start_pos = data.find(start_tag);
            size_t end_pos = data.find(end_tag);
            if (start_pos != std::string::npos && end_pos != std::string::npos)
            {
                start_pos += start_tag.length();
                return data.substr(start_pos, end_pos - start_pos);
            }
            return "";
        };

        (*gps_data)["lat"] = parse_tag_value("lat");
        (*gps_data)["lon"] = parse_tag_value("lon");
        (*gps_data)["time"] = parse_tag_value("time");
        (*gps_data)["speed"] = parse_tag_value("speed");
        (*gps_data)["pdop"] = parse_tag_value("pdop");
        (*gps_data)["hdop"] = parse_tag_value("hdop");

        return gps_data;
    }

    std::shared_ptr<std::unordered_map<std::string, std::string>> parse_ahrs_data(const std::string& data)
    {
        auto ahrs_data = std::make_shared<std::unordered_map<std::string, std::string>>();

        auto parse_tag_value = [&data](const std::string& tag) -> std::string {
            std::string start_tag = "<" + tag + ">";
            std::string end_tag = "</" + tag + ">";
            size_t start_pos = data.find(start_tag);
            size_t end_pos = data.find(end_tag);
            if (start_pos != std::string::npos && end_pos != std::string::npos)
            {
                start_pos += start_tag.length();
                return data.substr(start_pos, end_pos - start_pos);
            }
            return "";
        };

        (*ahrs_data)["acc_x"] = parse_tag_value("x");
        (*ahrs_data)["acc_y"] = parse_tag_value("y");
        (*ahrs_data)["acc_z"] = parse_tag_value("z");
        (*ahrs_data)["gyro_x"] = parse_tag_value("x");
        (*ahrs_data)["gyro_y"] = parse_tag_value("y");
        (*ahrs_data)["gyro_z"] = parse_tag_value("z");
        (*ahrs_data)["mag_x"] = parse_tag_value("x");
        (*ahrs_data)["mag_y"] = parse_tag_value("y");
        (*ahrs_data)["mag_z"] = parse_tag_value("z");
        (*ahrs_data)["altitude"] = parse_tag_value("altitude");

        return ahrs_data;
    }

    double string_to_double(const std::string& str)
    {
        try
        {
            return std::stod(str);
        }
        catch (...)
        {
            return NAN;
        }
    }

    rclcpp::Publisher<drone_interfaces::msg::SensorData>::SharedPtr sensor_publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    serial::Serial serial_;
    std::string serial_port_;
    uint32_t baud_rate_;
};

int main(int argc, char *argv[])
{   
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SerialReader>());
    rclcpp::shutdown();
    return 0;
}

File: src/drone/scripts/GPS.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
import serial
from drone.msg import GPSData
import math

class SerialReader(Node):
    def __init__(self):
        super().__init__('serial_reader')
        self.publisher_ = self.create_publisher(GPSData, 'esp32_data', 10)
        self.timer = self.create_timer(0.1, self.timer_callback)
        
        self.serial_port = '/dev/ttyUSB0'  # Adjust to your serial port
        self.baud_rate = 115200  # Adjust to your baud rate
        self.ser = serial.Serial(self.serial_port, self.baud_rate, timeout=1)
        
        self.get_logger().info(f'SerialReader node started, reading from {self.serial_port} at {self.baud_rate} baud rate.')

    def timer_callback(self):
        if self.ser.in_waiting > 0:
            try:
                line = self.ser.readline().decode('utf-8').strip()
                self.get_logger().info(f'Read from serial: {line}')
                
                # Parse the GPS data from the string
                if line.startswith('<gps>') and line.endswith('</gps>'):
                    data = self.parse_gps_data(line)
                    if data:
                        msg = GPSData()
                        msg.lat = data.get('lat', float('nan'))  # Use NaN for missing latitude
                        msg.lon = data.get('lon', float('nan'))  # Use NaN for missing longitude
                        msg.date = data.get('date', 'N/A')  # Use 'N/A' for missing date
                        msg.time = data.get('time', 'N/A')  # Use 'N/A' for missing time
                        msg.course = data.get('course', float('nan'))  # Use NaN for missing course
                        msg.speed = data.get('speed', float('nan'))  # Use NaN for missing speed
                        msg.pdop = data.get('pdop', float('nan'))  # Use NaN for missing pdop
                        msg.hdop = data.get('hdop', float('nan'))  # Use NaN for missing hdop
                        self.publisher_.publish(msg)
            except Exception as e:
                self.get_logger().error(f'Error reading from serial port: {e}')

    def parse_gps_data(self, data):
        # Simple XML parsing
        try:
            gps_data = {}
            
            if '<lat>' in data and '</lat>' in data:
                gps_data['lat'] = float(data.split('<lat>')[1].split('</lat>')[0])
            else:
                gps_data['lat'] = float('nan')
            
            if '<lon>' in data and '</lon>' in data:
                gps_data['lon'] = float(data.split('<lon>')[1].split('</lon>')[0])
            else:
                gps_data['lon'] = float('nan')
            
            if '<date>' in data and '</date>' in data:
                gps_data['date'] = data.split('<date>')[1].split('</date>')[0]
            else:
                gps_data['date'] = 'N/A'
            
            if '<time>' in data and '</time>' in data:
                gps_data['time'] = data.split('<time>')[1].split('</time>')[0]
            else:
                gps_data['time'] = 'N/A'
            
            if '<course>' in data and '</course>' in data:
                gps_data['course'] = float(data.split('<course>')[1].split('</course>')[0])
            else:
                gps_data['course'] = float('nan')
            
            if '<speed>' in data and '</speed>' in data:
                gps_data['speed'] = float(data.split('<speed>')[1].split('</speed>')[0])
            else:
                gps_data['speed'] = float('nan')
            
            if '<pdop>' in data and '</pdop>' in data:
                gps_data['pdop'] = float(data.split('<pdop>')[1].split('</pdop>')[0])
            else:
                gps_data['pdop'] = float('nan')
            
            if '<hdop>' in data and '</hdop>' in data:
                gps_data['hdop'] = float(data.split('<hdop>')[1].split('</hdop>')[0])
            else:
                gps_data['hdop'] = float('nan')
            
            return gps_data
        except Exception as e:
            self.get_logger().error(f'Error parsing GPS data: {e}')
            return None

def main(args=None):
    rclpy.init(args=args)
    serial_reader = SerialReader()
    rclpy.spin(serial_reader)
    serial_reader.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
